system/lib/libc/dynlink.c

```diff
@@ -223,23 +266,25 @@ static void ensure_init() {
 void* dlopen(const char* file, int flags) {
   ensure_init();
   if (!file) {
-    return head;
+    return head->dso;
   }
   dbg("dlopen: %s [%d]", file, flags);
 
-  struct dso* p;
   int cs;
   pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &cs);
   do_write_lock();
 #ifdef _REENTRANT
-  // Make sure we are in sync before loading any new DSOs.
+  // Make sure we are in sync before performing any write operations.
   dlsync();
 #endif
 
+  struct dso* p;
+
   /* Search for the name to see if it's already loaded */
-  for (p = head; p; p = p->next) {
-    if (!strcmp(p->name, file)) {
-      dbg("dlopen: already opened: %p", p);
+  for (struct dlevent* e = head; e; e = e->next) {
```

Comment:
Shouldn't we skip dlsym events (if only for efficiency)?

Reply:
Done.